### PR TITLE
[Model] Add support for Qwen-Image-Edit-2509

### PR DESCRIPTION
**Pull request by Zyann7**
## Motivation

SGLang Diffusion currently only supports Qwen-Image-Edit, while Qwen-Image-Edit-2509 adds multi-image input. This PR adapts the integration to support `Qwen-Image-Edit-2509` properly.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines/qwen_image.py [hunk 1]
```diff
@@ -192,4 +192,77 @@ def create_pipeline_stages(self, server_args: ServerArgs):
...
+            stage_name="denoising_stage",
+            stage=DenoisingStage(
+                transformer=self.get_module("transformer"),
+                scheduler=self.get_module("scheduler"),
+            ),
+        )
+
+        self.add_stage(
+            stage_name="decoding_stage", stage=DecodingStage(vae=self.get_module("vae"))
+        )
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/pipelines/qwen_image.py]: could you apply this?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py [hunk 2]
```diff
@@ -244,80 +292,199 @@ def forward(
...
+                    image_latents = _pack_latents(
+                        image_latents,
+                        batch_size,
+                        num_channels_latents,
+                        image_latent_height,
+                        image_latent_width,
+                    )
+                    all_image_latents.append(image_latents)
+
+                batch.image_latent = torch.cat(all_image_latents, dim=1)
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py]: could you apply this?

**Review by mickqian — CHANGES_REQUESTED** [review]
Overall it seems good, but there are issue regarding code quality:
1. avoid code duplication
2. try to do more abstraction, especially avoid `instanceof`, which is intended to be temporary

Also, could you add it to server test?

Also, it looks like sp is not supported. You could refer to `QwenImagePipelineConfig` for help

**Comment by Zyann7** [comment]
> this pr sounds great and is of our highest priority. But could we reuse the previous logic as much as possible, to reduce duplication?

Sorry for the late reply! I’ve already submitted a commit with the updates. If further optimizations are needed, I’ll continue working on it.

File: python/sglang/multimodal_gen/configs/pipeline_configs/qwen_image.py [hunk 3]
```diff
@@ -285,3 +288,90 @@ def slice_noise_pred(self, noise, latents):
         # remove noise over input image
         noise = noise[:, : latents.size(1)]
         return noise
+
+
+class QwenImageEditPlusPipelineConfig(QwenImageEditPipelineConfig):
+    task_type: ModelTaskType = ModelTaskType.I2I
+
+    def _prepare_edit_cond_kwargs(
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/pipeline_configs/qwen_image.py]: a lot of duplicate code with `QwenImageEditPipelineConfig::_prepare_edit_cond_kwargs`

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py [hunk 4]
```diff
@@ -111,18 +115,42 @@ def forward(
...
-        ):
-            prompt_template_encode = "<|im_start|>system\nDescribe the key features of the input image (color, shape, size, texture, objects, background), then explain how the user's text instruction should alter or modify the image. Generate a new image that meets the user's requirements while maintaining consistency with the original input where appropriate.<|im_end|>\n<|im_start|>user\n<|vision_start|><|image_pad|><|vision_end|>{}<|im_end|>\n<|im_start|>assistant\n"
-            txt = prompt_template_encode.format(batch.prompt)
-            image_processor_kwargs = dict(text=[txt], padding=True)
+        prompt_image = server_args.pipeline_config.preprocess_image(
+            image, self.vae_image_processor
+        )
+
+        if batch.prompt:
+            if isinstance(server_args.pipeline_config, QwenImageEditPlusPipelineConfig):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py]: could we abstract these, into a method, something like `prepare_kwargs_for_image_processor`, of `PipelineConfig`?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py [hunk 5]
```diff
@@ -399,7 +389,8 @@ def forward(
 
         batch.image_latent = image_latents
 
-        self.maybe_free_model_hooks()
+        if hasattr(self, "maybe_free_model_hooks"):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py]: why do we change this?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py [hunk 6]
```diff
@@ -451,3 +442,121 @@ def verify_output(self, batch: Req, server_args: ServerArgs) -> VerificationResu
...
+
+        if isinstance(scaling_factor, torch.Tensor):
+            scaling_factor = scaling_factor.to(latent_condition.device)
+
+        latent_condition -= shift_factor
+        latent_condition = latent_condition * scaling_factor
+
+        return latent_condition
+
+    def _duplicate_and_pack_latents(
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py]: if this is `qwen-image-edit` specific, don't implement it here, in common logic

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/multimodal_gen/configs/pipeline_configs/qwen_image.py (hunk 3 above)
Comment (Zyann7) [reply python/sglang/multimodal_gen/configs/pipeline_configs/qwen_image.py]: The only difference was in how image_sizes were handled, I’ve extracted that part into a `_get_condition_image_sizes` helper and reused the common logic.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py (hunk 4 above)
Comment (Zyann7) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py]: done

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py (hunk 6 above)
Comment (Zyann7) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py]: done

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py (hunk 5 above)
Comment (Zyann7) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/image_encoding.py]: I’ve cleaned it up and reverted the code.

**Review by mickqian — APPROVED** [review]
Great, we should add it to CI in this PR, or a following PR, regarding multi-images input

**Comment by yhyang201** [comment]
Have you run your PR? Does it work properly with Qwen/Qwen-Image-Edit and Qwen/Qwen-Image-Edit-2509 (single-image input only)?
